File: Arbitration/MPArbitration/Model/AuthorityDisputeFee.cs
using System.Text.Json.Serialization;
using System.ComponentModel.DataAnnotations.Schema;
using Microsoft.EntityFrameworkCore;
using System.ComponentModel.DataAnnotations;
using System.Security.Permissions;

namespace MPArbitration.Model
{
    public interface IDisputeFee
    {
        int BaseFeeId { get; set; }
        BaseFee? BaseFee { get; set; }  // aka the fee "definition" containing the rules for calculation, etc
        FeeRecipient FeeRecipient { get; set; }
    }

    public class FeeStorageBase
    {
        [JsonPropertyName("amountDue")]
        public double AmountDue { get; set; }

        [JsonPropertyName("dueOn")]
        public DateTime? DueOn { get; set; } = null;

        [JsonPropertyName("feeRecipient")]
        [Column(TypeName = "nvarchar(20)")]
        public FeeRecipient FeeRecipient { get; set; } = FeeRecipient.Arbitrator;

        [JsonPropertyName("invoiceLink")]
        [StringLength(512)]
        public string InvoiceLink { get; set; } = "";

        [JsonPropertyName("invoiceReceivedOn")]
        public DateTime? InvoiceReceivedOn { get; set; } = null;

        [JsonPropertyName("isRefundable")]
        public bool IsRefundable { get; set; }

        [JsonPropertyName("isRequired")]
        public bool IsRequired { get; set; }

        [JsonPropertyName("paidBy")]
        [StringLength(60)]
        public string PaidBy { get; set; } = "";

        [JsonPropertyName("paidOn")]
        public DateTime? PaidOn { get; set; } = null;

        [JsonPropertyName("paymentMethod")]
        [StringLength(60)]
        public string PaymentMethod { get; set; } = "";

        [JsonPropertyName("paymentReferenceNumber")]
        [StringLength(255)]
        public string PaymentReferenceNumber { get; set; } = "";

        [JsonPropertyName("paymentRequestedOn")]
        public DateTime? PaymentRequestedOn { get; set; } = null;

        [JsonPropertyName("refundableAmount")]
        public double RefundableAmount { get; set; }

        [JsonPropertyName("refundAmount")]
        public double RefundAmount { get; set; }

        [JsonPropertyName("refundDueOn")]
        public DateTime? RefundDueOn { get; set; } = null;

        [JsonPropertyName("refundedOn")]
        public DateTime? RefundedOn { get; set; } = null;

        [JsonPropertyName("refundedTo")]
        [StringLength(60)]
        public string RefundedTo { get; set; } = "";

        [JsonPropertyName("refundMethod")]
        [StringLength(60)]
        public string RefundMethod { get; set; } = "";

        [JsonPropertyName("refundReferenceNumber")]
        [StringLength(255)]
        public string RefundReferenceNumber { get; set; } = "";

        [JsonPropertyName("refundRequestedBy")]
        [StringLength(60)]
        public string RefundRequestedBy { get; set; } = "";

        [JsonPropertyName("refundRequestedOn")]
        public DateTime? RefundRequestedOn { get; set; } = null;

        [JsonPropertyName("wasRefunded")]
        public bool WasRefunded { get; set; }

        [JsonPropertyName("wasRefundRequested")]
        public bool WasRefundRequested { get; set; }

    }

    /// <summary>
    /// Class used for importing CSV records using alternate foreign key values
    /// </summary>
    public class AuthorityDisputeFeeCSV : FeeStorageBase
    {
        /// <summary>
        /// If relevant, the text name of the Arbitration Entity that owns this Fee depending on the value of FeeRecipient.
        /// Without this property, fees imports would require a lot of duplicate logic.
        /// </summary>
        [JsonPropertyName("arbitratorEntityName")]
        public string ArbitratorEntityName { get; set; } = "";

        /// <summary>
        /// Combine with AuthorityCaseId to reference the targeted AuthorityDispute without using a database Id.
        /// </summary>
        [JsonPropertyName("authorityKey")]
        [Required]
        [StringLength(3)]
        public string AuthorityKey { get; set; } = "";

        /// <summary>
        /// Combine with AuthorityId to reference the targeted AuthorityDispute without using a database Id.
        /// </summary>
        [Required]
        [JsonPropertyName("authorityCaseId")]
        [StringLength(30)]
        public string AuthorityCaseId { get; set; } = "";  // aka "Dispute Number"

        /// <summary>
        /// Alternate primary key for locating the base fee
        /// </summary>
        [Required]
        [JsonPropertyName("feeName")]
        [StringLength(60)]
        public string FeeName { get; set; } = "";
    }

    /// <summary>
    /// Storage for Arbitrator and Authority fee instances
    /// </summary>
    [Index(nameof(AuthorityDisputeId),nameof(FeeRecipient),nameof(BaseFeeId),IsUnique = true)]
    public class AuthorityDisputeFee : FeeStorageBase, IDisputeFee
    {
        [JsonPropertyName("id")]
        public int Id { get; set; } = 0;

        [JsonPropertyName("baseFeeId")]
        public int BaseFeeId { get; set; } = 0;

        [NotMapped]
        [JsonPropertyName("baseFee")]
        public BaseFee? BaseFee { get; set; } = null;

        [JsonPropertyName("authorityDisputeId")]
        [Required]
        public int AuthorityDisputeId { get; set; } = 0;

        [JsonPropertyName("updatedBy")]
        [StringLength(60)]
        public string UpdatedBy { get; set; } = "";

        [JsonPropertyName("updatedOn")]
        public DateTime? UpdatedOn { get; set; } = null;
    }
}

File: Arbitration/MPArbitration/ClientApp/src/app/model/authority-dispute-fee.ts
import { LogLevel } from "@azure/msal-browser";
import { UtilService } from "../services/util.service";
import { loggerCallback } from "../app.module";
import { BaseFee } from "./base-fee";

export enum FeeRecipient {
    Arbitrator,
    Authority
}

export class AuthorityDisputeFee {
    public id = 0;
    public amountDue = 0;
    public authorityDisputeId = 0;
    public baseFee:BaseFee|undefined;
    public baseFeeId = 0;
    public createdBy ='';
    //public createdOn:Date|undefined;
    public dueOn:Date|undefined;
    public feeRecipient = FeeRecipient.Arbitrator;
    public invoiceLink = '';
    public invoiceReceivedOn:Date|undefined;
    public isExpanded = false;
    public isRefundable = false;
    public isRequired = false;
    public paidBy = '';
    public paidOn:Date|undefined;
    public paymentMethod = '';
    public paymentReferenceNumber = '';
    public paymentRequestedOn:Date|undefined;
    public refundableAmount = 0;
    public refundAmount = 0;
    public refundDueOn:Date|undefined;
    public refundedOn:Date|undefined;
    public refundedTo='';
    public refundMethod = '';
    public refundReferenceNumber='';
    public refundRequestedBy='';
    public refundRequestedOn:Date|undefined;
    public updatedBy = '';
    public updatedOn:Date|undefined;
    public wasRefunded = false;
    public wasRefundRequested = false;

    public get dueOnForPicker() {
        return !!this.dueOn ? this.dueOn.toLocaleDateString() : undefined;
    }
    public set dueOnForPicker(value:any){
        if(!value)
            this.dueOn = undefined;
        if(UtilService.IsValidUSDate(value))
            this.dueOn = new Date(value);
    }

    public get invoiceReceivedOnForPicker() {
        return !!this.invoiceReceivedOn ? this.invoiceReceivedOn.toLocaleDateString() : undefined;
    }
    public set invoiceReceivedOnForPicker(value:any){
        if(!value)
            this.invoiceReceivedOn = undefined;
        if(UtilService.IsValidUSDate(value))
            this.invoiceReceivedOn = new Date(value);
    }

    public get feeName():string {
        return !!this.baseFee ? this.baseFee.feeName : '';
    }
    public get paidByShort():string{
        if(!this.paidBy)
            return '';
        return UtilService.ToTitleCase(this.paidBy.substring(0,this.paidBy.indexOf('@')).replaceAll('.',' '));
    }

    public get paidOnForPicker() {
        return !!this.paidOn ? this.paidOn.toLocaleDateString() : undefined;
    }
    public set paidOnForPicker(value:any){
        if(!value)
            this.paidOn = undefined;
        if(UtilService.IsValidUSDate(value))
            this.paidOn = new Date(value);
    }

    public get paymentRequestedOnForPicker() {
        return !!this.paymentRequestedOn ? this.paymentRequestedOn.toLocaleDateString() : undefined;
    }
    public set paymentRequestedOnForPicker(value:any){
        if(!value)
            this.paymentRequestedOn = undefined;
        if(UtilService.IsValidUSDate(value))
            this.paymentRequestedOn = new Date(value);
    }

    public get refundDueOnForPicker() {
        return !!this.refundDueOn ? this.refundDueOn.toLocaleDateString() : undefined;
    }
    public set refundDueOnForPicker(value:any){
        if(!value)
            this.refundDueOn = undefined;
        if(UtilService.IsValidUSDate(value))
            this.refundDueOn = new Date(value);
    }

    public get refundedOnForPicker() {
        return !!this.refundedOn ? this.refundedOn.toLocaleDateString() : undefined;
    }
    public set refundedOnForPicker(value:any){
        if(!value)
            this.refundedOn = undefined;
        if(UtilService.IsValidUSDate(value))
            this.refundedOn = new Date(value);
    }

    public get refundRequestedOnForPicker() {
        return !!this.refundRequestedOn ? this.refundRequestedOn.toLocaleDateString() : undefined;
    }
    public set refundRequestedOnForPicker(value:any){
        if(!value)
            this.refundRequestedOn = undefined;
        if(UtilService.IsValidUSDate(value))
            this.refundRequestedOn = new Date(value);
    }

    constructor(obj?:any) {
        if(!obj)
            return;
    
        Object.assign(this, JSON.parse(JSON.stringify(obj)));
        try {
            if(this.invoiceReceivedOn)
                this.invoiceReceivedOn = new Date(obj.invoiceReceivedOn);
            if(this.dueOn)
                this.dueOn = new Date(obj.dueOn); 
            if(this.paidOn)
                this.paidOn = new Date(obj.paidOn);
            if(this.paymentRequestedOn)
                this.paymentRequestedOn = new Date(obj.paymentRequestedOn);
            if(this.refundDueOn)
                this.refundDueOn = new Date(obj.refundDueOn);
            if(this.refundedOn)
                this.refundedOn = new Date(obj.refundedOn);
            if(this.refundRequestedOn)
                this.refundRequestedOn = new Date(obj.refundRequestedOn);
            if(this.updatedOn)
                this.updatedOn = new Date(obj.updatedOn);

            // objects
            if(!!obj.baseFee){
                this.baseFee = new BaseFee(obj.baseFee);
                this.baseFeeId = this.baseFee.id;
            }

        } catch(err) {
            loggerCallback(LogLevel.Error, 'Unable to instantiate Entity');
            console.error(err);
        }
    }
}
